### PR TITLE
[smartthings] SmartThings Binding initial contribution

### LINKED ISSUE #1694: [smartthings] Add SmartThings Binding
Hi:

I am interested in working on a Samsung Smartthings binding. has anyone else put any effort into it? I have a smartthings hub and a couple of devices that work with it. I am just completing an OpenHAB 2 binding for the cm11a so I am comfortable with binding development. I have read much of the smartthings developer documentation. I am struggling with how the OAuth integration works but I just connected IFTTT to both smartthings and OpenHAB so hopefully I can figure the rest of that out.

Do you think this would be a good addition to OpenHAB?

Are other interested in participating?

Thanks,

Bob Raker

### PR DESCRIPTION
**Pull request by BobRak**
This is my initial contribution of a binding for the Samsung Smartthings  hub. It has been in development since Jan 2017. It has been available in my [repo ](https://github.com/BobRak/OpenHAB-Smartthings) for more than a year now and has been tested by a number of users. Smartthings defines a set of "capabilities" that every device must conform to. This binding supports every Smartthings capability. Many have been tested by me or others who have downloaded the binding. But, not all capabilities have been tested.

Signed-off-by: Bob Raker <rjraker@gmail.com> (github: bobrak)

Fixes #1694

### CODE REVIEW

**Comment by BobRak** [comment]
Question... There are 2 files in src/main/smartthings/.. that the user of this binding has to copy to their smartthings environment. Once the binding has been incorporated into openHAB is there a place where these files can be stored in openHAB and be made available for the user to download? Smartthings also has a way for a developer to submit open source modules for storage on their platform. I have submitted a request for them to be stored on the Smartthings platform I'll update this PR when I have news from them. If Smartthings is agreeable to hosting these files I'll update the installation instructions.

File: addons/binding/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/config/SmartthingsBridgeConfig.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+ */
+package org.openhab.binding.smartthings.config;
+
+/**
+ * Configuration data for Smartthings hub
+ *
+ * @author Bob Raker - Initial contribution
+ *
+ */
+
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/config/SmartthingsBridgeConfig.java]: Please remove newlines between javadoc and classes

File: addons/binding/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/config/SmartthingsBridgeConfig.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.smartthings.config;
+
+/**
+ * Configuration data for Smartthings hub
+ *
+ * @author Bob Raker - Initial contribution
+ *
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/config/SmartthingsBridgeConfig.java]: Please remove newlines below the author tag

File: addons/binding/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/discovery/SmartthingsDeviceData.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+/**
+ * Mapping object for data returned from smartthings hub
+ *
+ * @author Bob Raker - Initial contribution
+ *
+ */
+
+public class SmartthingsDeviceData {
+    private String capability;
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/discovery/SmartthingsDeviceData.java]: It if is a DTO (data transfer object), please make all fields public and remove getters and the empty constructor.
Also move dto's into an own java package. Most bindings call that "model" or "dto".

File: addons/binding/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/discovery/SmartthingsDiscoveryService.java [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+ *
+ */
+public class SmartthingsDiscoveryService extends AbstractDiscoveryService implements EventHandler {
+
+    private static final int SEARCH_TIME = 30;
+    private static final int INITIAL_DELAY = 5;
+    private static final int SCAN_INTERVAL = 180;
+
+    private final Pattern findIllegalChars = Pattern.compile("[^A-Za-z0-9_-]");
+
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/discovery/SmartthingsDiscoveryService.java]: You love newlines, I see. But also think about us reviewers. You binding is 5000 LOC, and I bet 500 lines are "unnecessary" (subjective) newlines.

If possible can you group fields and remove newlines between single entries? Usually final fields are one group, one-time initialized and runtime fields are groups.

File: addons/binding/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/discovery/SmartthingsDiscoveryService.java [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+    private ScheduledFuture<?> scanningJob;
+
+    public SmartthingsDiscoveryService(SmartthingsBridgeHandler bridgeHandler) {
+        super(SmartthingsBindingConstants.SUPPORTED_THING_TYPES_UIDS, SEARCH_TIME);
+        logger.debug("Initializing discovery service");
+        this.bridgeHandler = bridgeHandler;
+
+        // Get a Gson instance
+        gson = new Gson();
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/discovery/SmartthingsDiscoveryService.java]: Move init to the field and remove the superfluous comment

File: addons/binding/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/discovery/SmartthingsDiscoveryService.java [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+
+import com.google.gson.Gson;
+
+/**
+ * Smartthings Discovery service
+ *
+ * @author Bob Raker - Initial contribution
+ *
+ */
+public class SmartthingsDiscoveryService extends AbstractDiscoveryService implements EventHandler {
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/discovery/SmartthingsDiscoveryService.java]: Please make this a `@Component`, immediate=false

Implement `ThingHandlerService` here and return the class in `getServices()` in the `SmartthingsBridgeHandler`.

File: addons/binding/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/discovery/SmartthingsDiscoveryService.java [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+    }
+
+    public SmartthingsDiscoveryService() {
+        super(SmartthingsBindingConstants.SUPPORTED_THING_TYPES_UIDS, SEARCH_TIME);
+        logger.debug("Initializing discovery service with default constructor");
+    }
+
+    @Override
+    protected void activate(Map<String, Object> config) {
+        logger.debug("SmartthingsDiscoveryService.activate() called");
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/discovery/SmartthingsDiscoveryService.java]: Please use the OSGi console for checking if a service is running and don't implement activate/modifed/deactivate only for logging

File: addons/binding/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/discovery/SmartthingsDiscoveryService.java [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+     * Scanning worker class.
+     */
+    public class SmartthingsScan implements Runnable {
+        /**
+         * Handler for delegation to callbacks.
+         */
+        private SmartthingsDiscoveryService service;
+
+        /**
+         * Constructor.
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/discovery/SmartthingsDiscoveryService.java]: If you don't have to say anything in a javadoc, leave it out. Think about your LOC's.

File: addons/binding/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/handler/SmartthingsAbstractHandler.java [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+ * This is an abstract base class for the "Thing" handlers (i.e. SmartthingsXxxxeHandler).
+ * It is not used the the Bridge handler (SmartthingsBridgeHandler)
+ *
+ * @author Bob Raker - Initial contribution
+ *
+ */
+public abstract class SmartthingsAbstractHandler extends BaseThingHandler {
+
+    /**
+     * The device name that corresponds to the name in the Smartthings Hub
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/handler/SmartthingsAbstractHandler.java]: The field name is self-explaining. Please remove the comments and newlines between the fields

File: addons/binding/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/handler/SmartthingsBridgeHandler.java [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+    @Override
+    public void dispose() {
+        logger.debug("Smartthings Handler disposed.");
+
+        httpClient.stopHttpClient();
+
+        super.dispose();
+    }
+
+    private boolean validateConfig(SmartthingsBridgeConfig config) {
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/handler/SmartthingsBridgeHandler.java]: Extend from `ConfigStatusThingHandler`. That allows you to validate and return a more explicit error to the framework and user-interfaces.

File: addons/binding/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/handler/SmartthingsThingHandler.java [hunk 11]
```diff
@@ -0,0 +1,331 @@
...
+
+    private Logger logger = LoggerFactory.getLogger(SmartthingsThingHandler.class);
+
+    private SmartthingsThingConfig config;
+    private String smartthingsName;
+
+    private Map<ChannelUID, SmartthingsConverter> converters = new HashMap<ChannelUID, SmartthingsConverter>();
+
+    /**
+     * The constructor
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/handler/SmartthingsThingHandler.java]: Please remove if possible

File: addons/binding/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/handler/SmartthingsThingHandler.java [hunk 12]
```diff
@@ -0,0 +1,331 @@
...
+                // The smartthings hub won't (can't) return a response to this call. But, it will send a separate
+                // message back to the SmartthingBridgeHandler.receivedPushMessage handler
+            }
+
+            try {
+                httpClient.sendDeviceCommand(path, jsonMsg);
+                // Smartthings will not return a response to this message but will send it's response message
+                // which will get picked up by the SmartthingBridgeHandler.receivedPushMessage handler
+            } catch (InterruptedException | TimeoutException | ExecutionException e) {
+                logger.info("Attempt to send command to the Smartthings hub failed with: {}", e);
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/handler/SmartthingsThingHandler.java]: Remove the last "{}". The coding guidelines require to add an exception as last argument, not part of the message. That way the logging framework not only calls "toString()" on the exception, but prints an actual stack trace.

If that is not what you want, you need to explicitly use exception.toString()

File: addons/binding/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/SmartthingsHandlerFactory.java [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+    private List<SmartthingsThingHandler> thingHandlers = new ArrayList<SmartthingsThingHandler>();
+
+    @Override
+    public boolean supportsThingType(ThingTypeUID thingTypeUID) {
+        return THING_TYPE_SMARTTHINGS.equals(thingTypeUID) || SUPPORTED_THING_TYPES_UIDS.contains(thingTypeUID);
+    }
+
+    public SmartthingsHandlerFactory() {
+        // Get a Gson instance
+        gson = new Gson();
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/SmartthingsHandlerFactory.java]: Ideally, you create only one gson instance for your binding here in the factory and hand it over to the thinghandlers.

File: addons/binding/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/SmartthingsHandlerFactory.java [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+    }
+
+    @Override
+    protected ThingHandler createHandler(Thing thing) {
+        ThingTypeUID thingTypeUID = thing.getThingTypeUID();
+        logger.debug("SmartthingsHandlerFactory is now processing ThingTypeUID {}", thingTypeUID.getAsString());
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/SmartthingsHandlerFactory.java]: The framework will already log the creation of a thing.

File: addons/binding/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/SmartthingsHandlerFactory.java [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+            }
+        }
+    }
+
+    /**
+     * Register a new discovery service
+     *
+     * @param handler
+     */
+    private void registerDeviceDiscoveryService(SmartthingsBridgeHandler handler) {
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/SmartthingsHandlerFactory.java]: As said above: Don't start discovery services in the factory. They are either thingHandler services or independent services.

File: addons/binding/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/SmartthingsHttpClient.java [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+    private String smartthingsIp;
+    private int smartthingsPort;
+
+    public SmartthingsHttpClient(String smartthingsIp, int smartthingsPort) throws Exception {
+        // Save the Smartthing IP and port
+        this.smartthingsIp = smartthingsIp;
+        this.smartthingsPort = smartthingsPort;
+
+        // Setup the jetty client for use in sending data to the Smartthings hub
+        httpClient = new HttpClient();
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/SmartthingsHttpClient.java]: That's not allowed as per coding guidelines.
You must instead "reference" the OSGi service `HttpClientFactory` (in your factory/discovery component or other components) and get a client from there. The reason is that the framework want a single ssl configuration and control about open ports.

File: addons/binding/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/SmartthingsServlet.java [hunk 17]
```diff
@@ -0,0 +1,174 @@
...
+import com.google.gson.Gson;
+
+/**
+ * Receives all Http data from the Smartthings Hub
+ *
+ * @author Bob Raker - Initial contribution
+ *
+ */
+@SuppressWarnings("serial")
+public class SmartthingsServlet extends HttpServlet {
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/SmartthingsServlet.java]: Please make this a `@Component`

File: addons/binding/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/SmartthingsStateData.java [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+ */
+package org.openhab.binding.smartthings.internal;
+
+/**
+ * Data object for smartthings state data
+ *
+ * @author Bob Raker - Initial contribution
+ *
+ */
+public class SmartthingsStateData {
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/SmartthingsStateData.java]: DTO?

File: addons/binding/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/converter/SmartthingsDefaultConverter.java [hunk 19]
```diff
@@ -0,0 +1,49 @@
...
+
+/**
+ * Base converter class.
+ * The converter classes are responsible for converting "state" messages from the smartthings hub into openHAB States.
+ * And, converting handler.handleCommand() into messages to be sent to smartthings
+ *
+ * @author Bob Raker - Initial contribution
+ *
+ */
+public class SmartthingsDefaultConverter extends SmartthingsConverter {
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/converter/SmartthingsDefaultConverter.java]: I don't understand the purpose of this class?! Just use `SmartthingsConverter` as a non abstract class.

File: addons/binding/org.openhab.binding.smartthings/src/main/smartthings/DeviceHandlers/OpenHabDeviceHandler.groovy [hunk 20]
```diff
@@ -0,0 +1,147 @@
+/**
+ *  OpenHAB Bridge
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.smartthings/src/main/smartthings/DeviceHandlers/OpenHabDeviceHandler.groovy]: groovy files under "src/main" are not allowed

File: addons/binding/org.openhab.binding.smartthings/src/main/smartthings/SmartApps/OpenHabAppV2.groovy [hunk 21]
```diff
@@ -0,0 +1,919 @@
+/**
+ *  OpenHabAppV2
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.smartthings/src/main/smartthings/SmartApps/OpenHabAppV2.groovy]: groovy files under "src/main" are not allowed

**Comment by davidgraeff** [comment]
> There are 2 files in src/main/smartthings/.. that the user of this binding has to copy to their smartthings environment.

You can have them in your binding directory and let them be hosted by github. Just not in "src/".

The code quality is overall quite good, but there are some new concepts like the `ConfigStatusProvider` and the `HttpClientFactory` that should be used. A discovery service should really be an OSGi service, managed by OSGi and not manually in the factory. 

If you adapt the code to those new concepts and maybe (hopefully) reduce some of the whitespace, we can merge this soon.

The very last step is the adaption to the new build-system though.

**Comment by J-N-K** [comment]
We recently decided to put "additional files" in `<binding directory>/contrib`.

**Comment by BobRak** [comment]
@davidgraeff 
Thanks for the feedback. I have a couple of questions.

**First**
Regarding your comment: 

> The code quality is overall quite good, but there are some new concepts like the ConfigStatusProvider and the HttpClientFactory that should be used. A discovery service should really be an OSGi service, managed by OSGi and not manually in the factory.

Can you point me to examples of these techniques? If so, I can incorporate those techniques.

**Second**
The binding has been on my github for more than a year and there have been a number of users and quite a bit of feedback I've incorporated. But, recently I discovered a bug in the code related to colored bulbs and am working on that. Unfortunately the Smartthings documentation is quite poor in that area so It is taking some time. Plus I have not owned any of these color bulbs but now have one on order just so I can test.

And, Smartthings publishes a **[Capability Reference](https://docs.smartthings.com/en/latest/capabilities-reference.html)** and I have duplicated those capabilities in this binding. I have recently discovered there have been a bunch of additions to it for appliances and I have had a couple of requests to support those. So, I'm going to do that before I resubmit this pull request

Thanks for your review.
Bob

**Comment by BobRak** [comment]
David: @davidgraeff 
I've made most of the OSGI related changes you requested but now it does not run correctly. I converted the DiscoveryService to ThingHandlerService. I created a SmartthingsDiscoveryService.xml and included immediate="false". For some reason 3 instances of the Discovery service get created right away. When I look in the console none of them show up. When I try to do a manual discovery scan the DiscoveryService runs but it seems to be an instance that hasn't had the setBridgeHandler method called and therefore it fails. I can imagine none of this makes sense to you without seeing the code. If I uploaded it into my github repo can you look at it and try to tell me what I'm doing wrong? **I've looked for similar examples in the addons repo and haven't found anything helpful Is there something you can point me to?**

Any suggestion you have for moving this forward will be appreciated. I've probably spent 15-20 hours on this in the last week. Overall I've been working on this binding for a couple of years as time has permitted. I've had a large number of people download it and provide feedback. I would really like to see it finished. The Samsung Smartthings hub is a popular hub and I think this would be a good addition to OpenHAb. But, I'm stuck.

**Comment by davidgraeff** [comment]
There is a bug in openhab core that forces us at the moment to perform very strict annotations.
Usually it should be enough what you have done, at the moment you need to do this though:

```java
@Component(service = ThingHandlerService.class)
@NonNullByDefault
public class ModbusThingHandlerDiscoveryService extends ThingHandlerService {

}
```

**Comment by BobRak** [comment]
@davidgraeff Thanks for the quick feedback. I'm out of town for the next week but will make this change as soon as I get home. I'll let you know.
Bob

**Comment by BobRak** [comment]
@davidgraeff 
David:
## Multiple instances created.
I made the changes you recommended to the Discovery Service but I am still getting multiple instances of both my SmartthingsDiscoveryService and SmartthingsHandlerFactory created. I have uploaded all of my code to my Repo. Could I trouble to have a look at it and see if you can identify why multiple instances are being created. You can see it [here](https://github.com/BobRak/OpenHAB-CodeReview).

Here is a listing of what happens at run time. The values (id=xxx) are just to keep track of which instance is being used. These numbers change with every restart (of course).

```
SmartthingsServlet (id=114)
    new instance created

SmartthingsHandlerFactory (id=246)
    new instance created
    setHttpClientFactory()

SmartthingsHandlerFactory (id=307)
    new instance created
    setHttpClientFactory()

SmartthingsDiscoveryService (id=317)
    new instance created
    setSmartthingsHandlerFactory()
    getSmartthingsBridgeHandler() (bridge is null at this time)
    activate()
    startBackgroundDiscovery()

SmartthingsDiscoveryService (id=403)
    new instance created
    setSmartthingsHandlerFactory()
    getSmartthingsBridgeHandler() (bridge is null at this time)
    activate()
    startBackgroundDiscovery()

SmartthingsHandlerFactory (id=307)
    createHandler()
        bridge createHandler

SmartthingsDiscoveryService (id=562)
    new instance created
    setThingHandler(0)
    activate()
    startBackgroundDiscovery()

getSmartthingsBridgeHandler
    initialize()

SmartthingsDiscoveryService (id=562)
    sendSmartthingsDiscoveryRequest()

SmartthingsDiscoveryService (id=403)
    handleEvent()
        This ends badly because this is not the fully initialized instance 

```

## Question about your earlier code review comment...

You made the following comment about my SmartthingsDiscoveryService:

> Implement ThingHandlerService here and return the class in getServices() in the SmartthingsBridgeHandler

getServices() returns **Collection<Class<? extends ThingHandlerService>>**. I can't for the life of me figure out the code to get the SmartthingsDiscoveryService out of it. When I run it in the debugger I see it is there but I can't figure out the java code. Sorry, I realize this is basic java but I'm can't figure it out. Can you tell me how to get that or point me to an example. Thanks.

## Comment about @Reference to SmartthingsHandlerFactory in SmartthingsDiscoveryService

I realize this isn't necessary if I can figure out how to get the SmartthingsDiscoveryService from the getServices() call. I'll remove that and start the background discovery from the bridge when I know if has been created. That will solve some some issues.

## Thanks

Thanks for your help. Hopefully I can get this finished up in the next couple of weeks.

**Comment by BobRak** [comment]
@davidgraeff
David:

## Multiple instances created.
I made the changes you recommended to the Discovery Service but I am still getting multiple instances of both my SmartthingsDiscoveryService created. I have uploaded all of my code to my Repo. Could I trouble to have a look at it and see if you can identify why multiple instances are being created. You can see it here.

Here is a listing of what happens at run time. The values (id=xxx) are just to keep track of which instance is being used. These numbers change with every restart (of course).

```
SmartthingsDiscoveryService (id=87)
    new instance created
    activate()

SmartthingsServlet (id=114)
    new instance created

SmartthingsHandlerFactory (id=246)
    new instance created
    setHttpClientFactory()

SmartthingsDiscoveryService (id=326)
    getSmartthingsBridgeHandler() (bridge is null at this time)
    activate()

SmartthingsHandlerFactory (id=269)
    createHandler()
        bridge createHandler
        SmarthingsBridgeHandler - new instance (id=519)
        The call to getServices() returns an empty Collection 

SmarrthingsBridgeHandler 
    validateConfig

```
I can't figure out why the SamrtthingsDiscoverService is getting created right at the beginning. immediate is set to false. I thought maybe it had something to do implementing EventHandler but even if I comment all of that out it is started right away. Ideas?

## Question about your earlier code review comment...

You made the following comment about my SmartthingsDiscoveryService:

> Implement ThingHandlerService here and return the class in getServices() in the SmartthingsBridgeHandler

It took me a long time to figure out what you wanted here but I think I finally have it. You can see it in SmarrthingsBridgeHandler at linje 64. Unfortunately the code isn't getting executed. Maybe because the DiscoveryService has already been created???

## How to proceed

At this point I'm not sure how to proceed since I am unable to control how and when the discovery service created and started. Unless you have a better suggestion I would like to go back to creating the DiscoveryService in code. Around line 130 in SmartthingsHandlerFactory. I need to get this finished soon. I'm going to be away much of May.

## Thanks
Thanks for your help. Hopefully I can get this finished up in the next couple of weeks.

**Comment by J-N-K** [comment]
What's the status here?

**Comment by BobRak** [comment]
I'm still working on it. I hope to submit it before the end of Feb 2020.

*Problems*
I made all of the changes requested. But this was at the time the build
process was changing and the Eclipse IDE was being changed to use BND. I
had lots of trouble getting a development environment working. After
setting up 6 different workspaces I finally gave up. But, I took it up
again in Nov and was able to get a development environment working
including the debug environment.

In addition to the IDE issues I couldn't get discovery to work as a
service. And, I saw some open issues around that. But, I have recently
gotten that to work.

*What's left*
Smartthings has added a bunch of new capabilities (i.e. devices) such as an
air conditioner, washer, dryer. I am working at adding those. I don't plan
to add all of them but will try to add a couple that have been requested.

Testing - I need to do more.

*Question about Eclipse*
I setup a new workspace in late Nov and it worked well. I was able to code.
The debug environment worked including setting breakpoints in my code and
examining variables when paused. The UI's (paperUI and basicUI) worked.
Then the UI's quit working. My code still runs and breakpoints still work.
And, I see the expected messages for my binding in the log. Of course
without the UI's it is impossible to do much testing. The only thing I can
think of is maybe I did a MVN build. Has anyone else experienced similar
problems and is there a solution? Otherwise I'll setup another workspace,
that would be #8.

Here is the output from the paperUI
[image: image.png]

Bob Raker
git ID BobRak

**Comment by BobRak** [comment]
Hi:
I'm ready to submit this binding again for review. I have incorporated all of the changes requested above. But, a lot has changed in the past year. Would it be acceptable to close this pull request and start a new one?

**Comment by J-N-K** [comment]
Closed in favor of #7550

**Pull request closed without merging**